support-figma/extended-layout-plugin/src/localization-module.ts

```diff
@@ -0,0 +1,329 @@
+/**
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Utils from "./utils";
+
+const STRING_RES_PLUGIN_DATA_KEY = "vsw-string-res";
+const CONSOLE_TAG = `${Utils.CONSOLE_TAG}-LOCALIZATION`;
+
+interface StringResource {
+  name: string;
+  translatable: boolean;
+  // It can be a simple string or a string array.
+  text: string | string[];
+  textNodes: string[];
+}
+
+export async function generateLocalizationData(uploadedStrings: string[]) {
+  Utils.log(CONSOLE_TAG, "Generate string res name");
+
+  await figma.loadAllPagesAsync();
+
+  let output = new Map<string, StringResource>();
+
+  // strings.xml files does not allow duplicates so no checks for duplicates here.
+  for (let uploadedString of uploadedStrings) {
+    let strRes = uploadedString as unknown as StringResource;
+    output.set(strRes.name, strRes);
+  }
+
+  for (let page of figma.root.children) {
+    for (let child of page.children) {
+      await localizeNodeAsync(child, output);
+    }
+  }
+
+  // Convert to an array of key-value pairs
+  const outputArray = Array.from(output);
+  figma.ui.postMessage({ msg: "localization-output", output: outputArray });
+}
+
+export async function updateStringResName(strRes: StringResource) {
+  Utils.log(CONSOLE_TAG, "User renamed, update string res name");
+  await figma.loadAllPagesAsync();
+
+  for (let nodeId of strRes.textNodes) {
+    const node = await figma.getNodeByIdAsync(nodeId);
+    if (node?.type == "TEXT") {
+      saveToPluginData(node, strRes.name);
+    } else {
+      Utils.error(
+        CONSOLE_TAG,
+        "%s not a text node to update string res name",
+        nodeId
+      );
+    }
+  }
+}
+
+export async function deleteStringResName(strRes: StringResource) {
```

Comment:
Is this function used?

Reply:
No. Removed it. I will add this when I add support to exclude a node from translations.